src/Features/Core/Portable/Completion/Providers/Snippets/AbstractSnippetCompletionProvider.cs

```diff
@@ -35,34 +47,115 @@ public override async Task<CompletionChange> GetChangeAsync(Document document, C
             var allTextChanges = await allChangesDocument.GetTextChangesAsync(document, cancellationToken).ConfigureAwait(false);
 
             var change = Utilities.Collapse(allChangesText, allTextChanges.AsImmutable());
-            return CompletionChange.Create(change, allTextChanges.AsImmutable(), newPosition: snippet.CursorPosition, includesCommitCharacter: true);
+            var finalTextChange = ExtendSnippetTextChange(change, snippet.Placeholders);
+            var lspSnippet = GenerateLSPSnippet(finalTextChange, snippet.Placeholders);
+            var props = ImmutableDictionary<string, string>.Empty
+                .Add("LSPSnippet", lspSnippet!);
+
+            return CompletionChange.Create(change, allTextChanges.AsImmutable(), properties: props, snippet.CursorPosition, includesCommitCharacter: true);
         }
 
-        public override async Task ProvideCompletionsAsync(CompletionContext context)
+        private static string? GenerateLSPSnippet(TextChange textChange, ImmutableArray<RoslynLSPSnippetItem> placeholders)
```

Comment:
i would move this to a helper class, and then just unit test the heck out of this code directly.